platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ScopedServiceRegistry.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.service;
+
+import org.gradle.api.NonNullApi;
+import org.gradle.internal.service.scopes.Scope;
+
+/**
+ * A registry validating that all registered services are annotated with a corresponding {@link Scope}.
+ */
+@NonNullApi
+public class ScopedServiceRegistry extends DefaultServiceRegistry {
```

Comment:
❓ Why not just add this functionality to `DefaultServiceRegistry` directly?  You could add some additional constructors that take the scope to validate, and if `null` is provided do nothing differently to maintain the old behavior.

Reply:
It could. The separate type is to keep the general purpose implementation decoupled from the implementation that is used only for the services of a scope. I think the goal would be to remove this specialized subtype and move the logic into `ServiceRegistryBuilder` once all of the scopes use it instead of extension to define their services.